File: src/sources/constructionModelesLassoRank.c
#include <stdlib.h>
#include <omp.h>
#include <gsl/gsl_linalg.h>
#include "EMGrank.h"
#include "utils.h"

// TODO: comment on constructionModelesLassoRank purpose
void constructionModelesLassoRank_core(
	// IN parameters
	const Real* pi,// parametre initial des proportions
	const Real* rho, // parametre initial de variance renormalisé
	int mini, // nombre minimal d'itérations dans l'algorithme EM
	int maxi, // nombre maximal d'itérations dans l'algorithme EM
	const Real* X,// régresseurs
	const Real* Y,// réponse
	Real tau, // seuil pour accepter la convergence
	const int* A1, // matrice des coefficients des parametres selectionnes
	int rangmin,	//rang minimum autorisé
	int rangmax,	//rang maximum autorisé
	// OUT parameters (all pointers, to be modified)
	Real* phi,// estimateur ainsi calculé par le Lasso
	Real* llh,// estimateur ainsi calculé par le Lasso
	// additional size parameters
	int n,// taille de l'echantillon
	int p,// nombre de covariables
	int m,// taille de Y (multivarié)
	int k,// nombre de composantes
	int L)// taille de glambda
{
	//On cherche les rangs possiblement intéressants
	int deltaRank = rangmax-rangmin+1;
	int Size = (int)pow(deltaRank,k);
	int* Rank = (int*)malloc(Size*k*sizeof(int));
	for (int r=0; r<k; r++)
	{
		// On veut le tableau de toutes les combinaisons de rangs possibles
		// Dans la première colonne : on répète (rangmax-rangmin)^(k-1) chaque chiffre :
		//   ça remplit la colonne
		// Dans la deuxieme : on répète (rangmax-rangmin)^(k-2) chaque chiffre,
		//   et on fait ça (rangmax-rangmin)^2 fois
		// ...
		// Dans la dernière, on répète chaque chiffre une fois,
		//   et on fait ça (rangmin-rangmax)^(k-1) fois.
		int indexInRank = 0;
		int value = 0;
		while (indexInRank < Size)
		{
			int upperBoundU = (int)pow(deltaRank,k-r-1);
			for (int u=0; u<upperBoundU; u++)
				Rank[mi(indexInRank++,r,Size,k)] = rangmin + value;
			value = (value+1) % deltaRank;
		}
	}

	//Initialize phi to zero, because unactive variables won't be assigned
	for (int i=0; i<p*m*k*L*Size; i++)
		phi[i] = 0.0;
	for (int i=0; i<L*Size*2; i++)
		llh[i] = INFINITY;

	//initiate parallel section
	int lambdaIndex;
	omp_set_num_threads(OMP_NUM_THREADS);
	#pragma omp parallel default(shared) private(lambdaIndex)
	{
	#pragma omp for schedule(dynamic,CHUNK_SIZE) nowait
	for (lambdaIndex=0; lambdaIndex<L; lambdaIndex++)
	{
		//On ne garde que les colonnes actives : active sera l'ensemble des variables informatives
		int* active = (int*)malloc(p*sizeof(int));
		int longueurActive = 0;
		for (int j=0; j<p; j++)
		{
			if (A1[mi(j,lambdaIndex,p,L)] != 0)
				active[longueurActive++] = A1[mi(j,lambdaIndex,p,L)] - 1;
		}
		if (longueurActive == 0)
			continue;

		//from now on, longueurActive > 0
		Real* phiLambda = (Real*)malloc(longueurActive*m*k*sizeof(Real));
		Real LLF;
		for (int j=0; j<Size; j++)
		{
			int* rank = (int*)malloc(k*sizeof(int));
			for (int r=0; r<k; r++)
				rank[r] = Rank[mi(j,r,Size,k)];
			Real* Xactive = (Real*)malloc(n*longueurActive*sizeof(Real));
			for (int i=0; i<n; i++)
			{
				for (int jj=0; jj<longueurActive; jj++)
					Xactive[mi(i,jj,n,longueurActive)] = X[mi(i,active[jj],n,p)];
			}
			Real* piLambda = (Real*)malloc(k*sizeof(Real));
			for (int r=0; r<k; r++)
				piLambda[r] = pi[mi(r,lambdaIndex,k,L)];
			Real* rhoLambda = (Real*)malloc(m*m*k*sizeof(Real));
			for (int u=0; u<m; u++)
			{
				for (int v=0; v<m; v++)
				{
					for (int r=0; r<k; r++)
						rhoLambda[ai(u,v,r,m,m,k)] = rho[ai4(u,v,r,lambdaIndex,m,m,k,L)];
				}
			}
			EMGrank_core(piLambda,rhoLambda,mini,maxi,Xactive,Y,tau,rank,
				phiLambda,&LLF,
				n,longueurActive,m,k);
			free(rank);
			free(Xactive);
			free(piLambda);
			free(rhoLambda);
			//llh[(lambdaIndex-1)*Size+j,] = c(LLF, ...)
			llh[mi(lambdaIndex*Size+j,0,L*Size,2)] = LLF;
			//sum(Rank[j,] * (length(active)- Rank[j,] + m)) )
			Real dotProduct = 0.0;
			for (int r=0; r<k; r++)
				dotProduct += Rank[mi(j,r,Size,k)] * (longueurActive-Rank[mi(j,r,Size,k)]+m);
			llh[mi(lambdaIndex*Size+j,1,Size*L,2)] = dotProduct;
			//phi[active,,,(lambdaIndex-1)*Size+j] = res$phi
			for (int jj=0; jj<longueurActive; jj++)
			{
				for (int mm=0; mm<m; mm++)
				{
					for (int r=0; r<k; r++)
					{
						phi[ai4(active[jj],mm,r,lambdaIndex*Size+j,p,m,k,L*Size)] =
							phiLambda[ai(jj,mm,r,longueurActive,m,k)];
					}
				}
			}
		}
		free(active);
		free(phiLambda);
	}
	}
	free(Rank);
}

File: pkg/R/constructionModelesLassoRank.R
#' constructionModelesLassoRank
#'
#' TODO: description
#'
#' @param ...
#'
#' @return ...
#'
#' export
constructionModelesLassoRank = function(pi, rho, mini, maxi, X, Y, tau, A1, rangmin,
	rangmax, ncores, verbose=FALSE)
{
  n = dim(X)[1]
  p = dim(X)[2]
  m = dim(rho)[2]
  k = dim(rho)[3]
  L = dim(A1)[2]

	# On cherche les rangs possiblement intéressants
  deltaRank = rangmax - rangmin + 1
  Size = deltaRank^k
  Rank = matrix(0, nrow=Size, ncol=k)
  for (r in 1:k)
	{
		# On veut le tableau de toutes les combinaisons de rangs possibles
		# Dans la première colonne : on répète (rangmax-rangmin)^(k-1) chaque chiffre :
		#   ça remplit la colonne
		# Dans la deuxieme : on répète (rangmax-rangmin)^(k-2) chaque chiffre,
		#   et on fait ça (rangmax-rangmin)^2 fois
		# ...
		# Dans la dernière, on répète chaque chiffre une fois,
		#   et on fait ça (rangmin-rangmax)^(k-1) fois.
    Rank[,r] = rangmin + rep(0:(deltaRank-1), deltaRank^(r-1), each=deltaRank^(k-r))
  }

  if (ncores > 1)
	{
    cl = parallel::makeCluster(ncores)
    parallel::clusterExport( cl, envir=environment(),
			varlist=c("A1","Size","Pi","Rho","mini","maxi","X","Y","tau",
			"Rank","m","phi","ncores","verbose") )
	}

	computeAtLambda <- function(lambdaIndex)
	{
		if (ncores > 1)
			require("valse") #workers start with an empty environment

    # on ne garde que les colonnes actives
    # 'active' sera l'ensemble des variables informatives
    active = A1[,lambdaIndex]
    active = active[-(active==0)]
		phi = array(0, dim=c(p,m,k,Size))
		llh = matrix(0, Size, 2) #log-likelihood
    if (length(active) > 0)
		{
      for (j in 1:Size)
			{
        res = EMGrank(Pi[,lambdaIndex], Rho[,,,lambdaIndex], mini, maxi,
					X[,active], Y, tau, Rank[j,])
        llh = rbind(llh,
					c( res$LLF, sum(Rank[j,] * (length(active)- Rank[j,] + m)) ) )
        phi[active,,,] = rbind(phi[active,,,], res$phi)
      }
    }
		list("llh"=llh, "phi"=phi)
	}

	#Pour chaque lambda de la grille, on calcule les coefficients
  out =
		if (ncores > 1)
			parLapply(cl, seq_along(glambda), computeAtLambda)
		else
			lapply(seq_along(glambda), computeAtLambda)

	if (ncores > 1)
    parallel::stopCluster(cl)

	# TODO: this is a bit ugly. Better use bigmemory and fill llh/phi in-place
	# (but this also adds a dependency...)
	llh <- do.call( rbind, lapply(out, function(model) model$llh) )
	phi <- do.call( rbind, lapply(out, function(model) model$phi) )
	list("llh"=llh, "phi"=phi)
}
